README.md
## 麒麟计算器

### 项目运行方式  
```bash
chmod 777 ./run  
./run kylin-calculator
```


### 更新历史

#### 20210611 09:08 cl
1.0.31kord15
解决键盘输入一位数，界面显示两位数的问题

#### 20210519 13:39 jdy
1.0.31kord13  
优化n次方/n次方根计算流程:禁止多次连续点击n次方/n次方根连续点击操作

#### 20210419 17:09 cl
1.0.31kord10
创建DBus服务，最小化后再次启动应用时拉起主界面

#### 20210419 16:38 jdy
1.0.31kord9  
添加新版日志  
修复一个debug模式发现的段错误  

#### 20210416 11:15 jdy
1.0.31kord8  
根据新版设计稿修改应用名称   

#### 20210401 11:46 cl
1.0.31kord7  
修复bug 34897 44911
实现需求 输入过长时字号变小，删除或清除时字号变大

#### 20210330 09:47 jdy
1.0.31kord6  
修复bug 45498 45469

#### 20210127 18:58 jdy
1.0.31

#### 20210126 19:21 cl
修复BUG 27685 27661 32782  
重构界面切换逻辑  
实现前端组件复用  
1.0.30 kord1~rc1

#### 20210125 21:54 jdy
移动翻译文件到translations  
删除二进制pm文件

#### 20210123 17:40 jdy
1.0.30

#### 20210116 09:27 jdy
1.0.29

#### 20210115 13:58 cl
1.0.28 Update Desktop

#### 20210114 19:55 cl
Fix bug 33043 汇率模式点击“00”无反应

#### 20210112 10:08 cl
1.0.27

#### 20210109 09:02 jdy
1.0.26

#### 20210108 15:55 jdy
1.0.25

#### 20210108 15:00 jdy
1.0.24

#### 20210107 10:45 cl
1.0.23  
增加菜单栏 (不完善，未添加gsetting)

#### 20201230 20:19 cl
1.0.23  
汇率模式及汇率列表适配主题颜色

#### 20201229 14:45 jdy
1.0.23

#### 20201229 10:23 cl
1.0.22  
适配主题颜色和三联按钮样式

#### 20201219 20:57 cl
1.0.21  
适配4K屏，标题栏按钮样式和三联保持一致

#### 20201214 20:57 jdy
1.0.20

#### 20201205 11:21 jdy
1.0.19

#### 20201204 20:28 jdy
1.0.18

#### 20201202 17:15 jdy
1.0.17~rc  
合并dev代码

#### 20201202 13:54 jdy
1.0.16版本

#### 20201127 15:37 jdy
1.0.15版本

#### 20201124 11:03 jdy
生成字符串处理md  
合并dev代码

#### 20201124 10:16 cl
生成ts和qm文件

#### 20201119 15:16 cl
适配desktop文件藏文翻译

#### 20201113 20:00 cl
汇率换算历史记录，ANS按钮

#### 20201111 17:18 cl
开发2nd按钮和Rad按钮功能，实现三角函数和反三角函数替换以及弧度角度替换

#### 20201111 17:25 jdy
合并dev分支代码 解决闪退bug

#### 20201107 15:53 jdy
合并dev分支代码

#### 20201107 15:01 cl
将汇率换算模式下的历史记录转换为相应的历史记录，使用unitConvHistory函数对原本的历史记录进行转换。

#### 20201107 09:58 cl
将计算器各个模式下的按钮替换为BasicButton，使用图片作为显示，重写setText和text函数，从而保证按钮显示不受系统字体影响。


debian/changelog
kylin-calculator (1.0.31kord17) v101; urgency=medium

  * BUG号：无
  * 任务号：43294 【EDU V2.0】【计算器】UI界面优化
  * 需求号：7227 【计算器】UI界面优化 
  * 其他改动说明：无

 -- jishengjie <jishengjie@kylinos.cn>  Mon, 09 Aug 2021 09:08:20 +0800

kylin-calculator (1.0.31kord16) v101; urgency=medium

  * BUG号：69563
  * 任务号：无.
  * 需求号：无.
  * 其他改动说明：
    - 【创新院】【计算器】使用计算器科学模式，输入“（—”后输入一个数字，应用自动闪退

 -- jiangdingyuan <jiangdingyuan@kylinos.cn>  Tue, 27 Jul 2021 09:08:20 +0800

kylin-calculator (1.0.31kord15) v101; urgency=medium

  * BUG号：60053, 60063.
  * 任务号：无.
  * 需求号：无.
  * 其他改动说明：
    - 解决键盘输入一位数，界面显示两位数的问题.

 -- caoliang <caoliang@kylinos.cn>  Fri, 11 Jun 2021 09:08:20 +0800

kylin-calculator (1.0.31kord14) v101; urgency=medium

  * BUG号：57818, 57815.
  * 任务号：无.
  * 需求号：无.
  * 其他改动说明：
    - 添加复制粘贴的翻译.

 -- caoliang <caoliang@kylinos.cn>  Wed, 26 May 2021 13:37:44 +0800

kylin-calculator (1.0.31kord13) v101; urgency=medium

  * BUG号：无.
  * 任务号：无.
  * 需求号：无.
  * 其他改动说明：
    - 优化n次方/n次方根计算流程:禁止多次连续点击n次方/n次方根连续点击操作

 -- jiangdingyuan <jiangdingyuan@kylinos.cn>  Wed, 19 May 2021 13:37:39 +0800

kylin-calculator (1.0.31kord12) v101; urgency=medium

  * BUG号：无.
  * 任务号：29433, 25409.
  * 需求号：无.
  * 其他改动说明：
    - 应用图标跟随图标主题变化.
    - 新增复制粘贴功能.

 -- caoliang <caoliang@kylinos.cn>  Fri, 14 May 2021 17:17:34 +0800

kylin-calculator (1.0.31kord11) v101; urgency=medium

  * BUG号：无.
  * 需求号：无.
  * 其他改动说明：
    - 对最小化后再次启动应用时拉起主界面的机制进行优化.
    - 根据规范修改changelog.
  * 影响域：无.

 -- caoliang <caoliang@kylinos.cn>  Fri, 23 Apr 2021 09:31:59 +0800

kylin-calculator (1.0.31kord10) v101; urgency=medium

  * BUG号：49255.
  * 需求号：无.
  * 其他改动说明：
    - 解决禅道bug#49255 : 打开窗体未置顶显示.
    - 创建DBus服务，最小化后再次启动应用时拉起主界面.

 -- caoliang <caoliang@kylinos.cn>  Mon, 19 Apr 2021 17:06:06 +0800

kylin-calculator (1.0.31kord9) v101; urgency=medium

  * BUG号：无.
  * 需求号：25395.
  * 其他改动说明：
    - 添加新版日志
    - 修复一个debug模式发现的段错误

 -- jiangdingyuan <jiangdingyuan@kylinos.cn>  Mon, 19 Apr 2021 16:36:14 +0800

kylin-calculator (1.0.31kord8) v101; urgency=medium

  * BUG号：无.
  * 需求号：26485.
  * 其他改动说明：
    - 根据新版设计稿修改应用名称
    - 麒麟计算器         --> 计算器
    - kylin-calculator --> Calculator
    - kylin calculator --> Calculator

 -- jiangdingyuan <jiangdingyuan@kylinos.cn>  Fri, 16 Apr 2021 10:57:24 +0800

kylin-calculator (1.0.31kord7) v101; urgency=medium

  * 解决禅道bug #34897 : 【麒麟计算器】汇率模式输入过多位数转换时，计算器窗口太小，导致内容显示不全.
  * 解决禅道bug #44911 : 【麒麟计算器】标准计算器计算结果有多位小数时显示不全.
  * 输入过长时字号变小，删除或清除时字号变大.

 -- caoliang <caoliang@kylinos.cn>  Thu, 01 Apr 2021 11:42:57 +0800

kylin-calculator (1.0.31kord6) v101; urgency=medium

  * 解决禅道bug # 45498 【麒麟计算器】当计算结果出现科学计数法时，连续点击退格键计算器闪退 
  * 解决禅道bug # 45469 【麒麟计算器】当计算结果出现科学计数法的时候，e的指数为10的倍数时计算结果错误 

 -- jiangdingyuan <jiangdingyuan@kylinos.cn>  Tue, 30 Mar 2021 09:47:08 +0800

kylin-calculator (1.0.31kord5) v101; urgency=medium

  * 解决禅道bug #43104 : 【麒麟计算器】点击关于按钮，软件闪退.
  * 添加打印日志文件.

 -- caoliang <caoliang@kylinos.cn>  Wed, 17 Mar 2021 14:35:17 +0800

kylin-calculator (1.0.31kord4) v101; urgency=medium

  * 解决禅道bug #40286 : 【麒麟计算器】计算汇率时删除部分输入的数值，汇率计算错误.
  * 对阶乘的错误计算进行处理.
  * 修复科学模式三角函数按钮和角度弧度按钮的显示逻辑.

 -- caoliang <caoliang@kylinos.cn>  Mon, 15 Mar 2021 09:47:47 +0800

kylin-calculator (1.0.31kord3) v101; urgency=medium

  * 解决禅道bug #38016 : 【计算器】输入小数时，小数位无法输入‘00’，点击就报错.
  * 解决禅道bug #42436 : 【麒麟计算器】科学模式下，小括号无法用键盘输入.
  * 修复货币选择框显示逻辑.
  * 更新汇率筛选逻辑.

 -- caoliang <caoliang@kylinos.cn>  Thu, 11 Mar 2021 19:49:03 +0800

kylin-calculator (1.0.31kord2) v101; urgency=medium

  * 解决禅道bug #40251 : 【麒麟计算器】英文模式下，汇率计算模式存在字体显示不全.
  * 界面背景色以及按钮悬浮效果适配主题框架.
  * 使用系统图标库中的应用图标.

 -- caoliang <caoliang@kylinos.cn>  Tue, 09 Mar 2021 13:40:40 +0800

kylin-calculator (1.0.31kord1) v101; urgency=medium

  * 解决禅道bug #40468 : 【麒麟计算器】使用汇率模式，切换夜间模式，再切换其他计算模式【科学】【标准】，键盘显示异常.
  * 解决禅道bug #40475 : 【麒麟计算器】汇率计算器，点击刷新按钮，没有任何回馈或者动效效果.
  * 解决禅道bug #41107 : 【麒麟计算器】关于中，邮箱前文字建议修改为“服务与支持团队”.

 -- caoliang <caoliang@kylinos.cn>  Thu, 04 Mar 2021 09:16:52 +0800

kylin-calculator (1.0.31kord) v101; urgency=medium

  [ Jiang Dingyuan ]
  * 关于窗口优化：添加模态，设置图标，添加超链接,添加悬浮提示
  * MainWindow添加单例
  * 删除翻译生成的qm文件 添加生成qm文件脚本
  * 实现需求 #4419：点击“=”后显示结果为数字不变，多次连续点击“=”只读取一次点击，其他视为无效点击。

 -- jiangdingyuan <jiangdingyuan@kylinos.cn>  Wed, 27 Jan 2021 18:47:16 +0800

kylin-calculator (1.0.30kord1~rc1) v101; urgency=medium

  * 解决禅道bug #27658 : 【麒麟计算器】汇率模式下货币切换导致历史记录显示异常.
  * 解决禅道bug #27661 : 【麒麟计算器】汇率模式未记住货币种类.
  * 解决禅道bug #32782 : 【麒麟计算器】模式切换后，历史记录显示异常.
  * 解决禅道bug #33471 : 【麒麟计算器】切换科学模式，未默认显示到屏幕中央位置.
  * 解决禅道bug #35353 : 【麒麟计算器】统一开始菜单图标和任务栏图标.

 -- caoliang <caoliang@kylinos.cn>  Tue, 26 Jan 2021 15:44:34 +0800

kylin-calculator (1.0.30kord) v101; urgency=medium

  * 解决禅道bug #32760【麒麟计算器】鼠标按住计算器窗口靠近边沿位置无法拖动窗口

 -- jiangdingyuan <jiangdingyuan@kylinos.cn>  Sat, 23 Jan 2021 17:39:11 +0800

kylin-calculator (1.0.29kord) v101; urgency=medium

  * 注释调试代码
  * 关于菜单细节调整

 -- jiangdingyuan <jiangdingyuan@kylinos.cn>  Sat, 16 Jan 2021 09:29:00 +0800

kylin-calculator (1.0.28kord) v101; urgency=medium

  * 解决禅道bug #33043 : 【麒麟计算器】汇率模式点击"00"没有反应.
  * 修改应用英文名和应用类别.

 -- caoliang <caoliang@kylinos.cn>  Fri, 15 Jan 2021 13:38:42 +0800

kylin-calculator (1.0.27kord) v101; urgency=medium

  * 解决禅道bug #31985 【麒麟计算器】科学计算模式下点击Rad后按钮变白
  * 解决禅道bug #32052 【麒麟计算器】汇率计算器界面显示异常
  * 解决禅道bug #32764 【麒麟计算器】深色模式下，汇率更新按钮显示不清晰
  * 解决禅道bug #32767 【麒麟计算器】货币子窗口不跟随主题颜色变色

 -- caoliang <caoliang@kylinos.cn>  Tue, 12 Jan 2021 10:08:50 +0800

kylin-calculator (1.0.26kord) v101; urgency=medium

  * 修复翻译导致的切换bug

 -- jiangdingyuan <jiangdingyuan@kylinos.cn>  Sat, 09 Jan 2021 09:04:27 +0800

kylin-calculator (1.0.25kord) v101; urgency=medium

  * 修复关于信息
  * 解决禅道bug #29272 【麒麟计算器】应用默认背景为黑色(深色）主题，与其他应用主题风格不协调

 -- jiangdingyuan <jiangdingyuan@kylinos.cn>  Fri, 08 Jan 2021 15:53:15 +0800

kylin-calculator (1.0.24kord) v101; urgency=medium

  * 修复界面bug， 适配黑色主题三联按钮

 -- jiangdingyuan <jiangdingyuan@kylinos.cn>  Fri, 08 Jan 2021 14:55:15 +0800

kylin-calculator (1.0.23kord) v101; urgency=medium

  *  解决禅道bug #30276 【麒麟计算器】输入小数点前等于或超过两位数的数字进行计算时结果显示“输入错误”


toolmodel.cpp
#include "toolmodel.h"

ToolModelOutput::ToolModelOutput(QWidget *parent)
    : QWidget(parent)
{
    // 汇率刷新
    toolLabUpdate  = new QLabel(this);
    toolIconUpdate = new QLabel(this);
    toolLabRate    = new QLabel(this);

    toolLabUpdate->setText("汇率更新");
    toolLabUpdate->setStyleSheet("font-family:SourceHanSansCN-ExtraLight;font-size:16px;color:#FFFFFF;");

    toolIconUpdate->setFixedSize(16, 16);
    toolIconUpdate->setStyleSheet("border-image:url(:/image/update.png);");

    toolLabRate->setText("2020.09.03 09:30\n1 CNY = 0.15 USD");
    toolLabRate->setStyleSheet("font-family:NotoSansHans-Light;font-size:10px;color:#A6A6A6;line-height:17px;");

    toolUpdateLayout = new QVBoxLayout(this);
    toolUpdateLayout->addWidget(toolLabUpdate,  0, Qt::AlignCenter);
    toolUpdateLayout->addWidget(toolIconUpdate, 0, Qt::AlignCenter);
    toolUpdateLayout->addWidget(toolLabRate,    0, Qt::AlignCenter);

    toolUpdateWid = new QWidget(this);
    toolUpdateWid->setLayout(toolUpdateLayout);
    toolUpdateWid->setFixedSize(106, 133);
//    toolUpdateWid->setFixedWidth(107);
    toolUpdateWid->setStyleSheet("background:#18181A;opacity:0.75;");

    // 换算前单位标识
    toolLabUnitBef = new QLabel(this);
    toolLabIconBef = new QLabel(this);

    QString unitLabStyle  = "font-family:SourceHanSansCN-ExtraLight;font-size:18px;color:#FFFFFF;line-height:22px;";

    toolLabUnitBef->setText("人民币\nCNY");
    toolLabUnitBef->setStyleSheet(unitLabStyle);

    toolLabIconBef->setFixedSize(10, 10);
    toolLabIconBef->setStyleSheet("border-image:url(:/image/downward.png);");

    toolUnitBefLayout = new QHBoxLayout(this);
    toolUnitBefLayout->addWidget(toolLabUnitBef);
    toolUnitBefLayout->addWidget(toolLabIconBef);

    toolUnitBefWid = new QWidget(this);
    toolUnitBefWid->setLayout(toolUnitBefLayout);
    toolUnitBefWid->setFixedSize(106, 66);
    toolUnitBefWid->setStyleSheet("background:#18181A;color:#FFFFFF;opacity:0.75;");

    // 换算后单位标识
    toolLabUnitAft = new QLabel(this);
    toolLabIconAft = new QLabel(this);

    toolLabUnitAft->setText("美元\nUSD");
    toolLabUnitAft->setStyleSheet(unitLabStyle);

    toolLabIconAft->setFixedSize(10, 10);
    toolLabIconAft->setStyleSheet("border-image:url(:/image/downward.png);");

    toolUnitAftLayout = new QHBoxLayout(this);
    toolUnitAftLayout->addWidget(toolLabUnitAft);
    toolUnitAftLayout->addWidget(toolLabIconAft);

    toolUnitAftWid = new QWidget(this);
    toolUnitAftWid->setLayout(toolUnitAftLayout);
    toolUnitAftWid->setFixedSize(106, 66);
    toolUnitAftWid->setStyleSheet("background:#18181A;color:#FFFFFF;opacity:0.75;");

    // 数据输出界面
    this->toolLabHis = new QLabel(this);
    this->toolLabBef = new QLabel(this);
    this->toolLabAft = new QLabel(this);

    toolLabHis->setWordWrap(true);
    toolLabBef->setWordWrap(true);
    toolLabAft->setWordWrap(true);

    QFont toolLabHisFont("SourceHanSansCN-Light", 40, 15);
    this->toolLabHis->setAlignment(Qt::AlignRight | Qt::AlignTop);
    this->toolLabHis->setFont(toolLabHisFont);
    this->toolLabHis->setText("");
    this->toolLabHis->setStyleSheet("color:#8C8C8C;font-size:20px;background-color:#18181A;");
    this->toolLabHis->setFixedHeight(133);
    this->toolLabHis->setContentsMargins(0, 0, 20, 0);
    this->toolLabHis->show();

    QFont toolLabBefFont("SourceHanSansCN-Normal;", 50, 15);
    this->toolLabBef->setAlignment(Qt::AlignRight | Qt::AlignBottom);
    this->toolLabBef->setFont(toolLabBefFont);
    this->toolLabBef->setText("0");
    this->toolLabBef->setStyleSheet("color:#FFFFFF;font-size:45px;background-color:#18181A;");
    this->toolLabBef->setFixedHeight(66);
    this->toolLabBef->setContentsMargins(0, 0, 20, 0);
    this->toolLabBef->show();

    QFont toolLabAftFont("SourceHanSansCN-Normal", 50, 15);
    this->toolLabAft->setAlignment(Qt::AlignRight | Qt::AlignBottom);
    this->toolLabAft->setFont(toolLabAftFont);
    this->toolLabAft->setText("0");
    this->toolLabAft->setStyleSheet("color:#FFFFFF;font-size:45px;background-color:#18181A;");
    this->toolLabAft->setFixedHeight(66);
    this->toolLabAft->setContentsMargins(0, 0, 20, 0);
    this->toolLabAft->show();

    // 数据单位列表
    this->unitListBef = new QListWidget(this);
    this->unitListAft  = new QListWidget(this);

    this->unitListBef->hide();
    this->unitListAft->hide();

    // 将数据输出组件和数据单位列表加入布局
    QGridLayout *toolOutputLayout = new QGridLayout(this);

    toolOutputLayout->addWidget(toolUpdateWid, 0, 0, 1, 1);
    toolOutputLayout->addWidget(toolUnitBefWid, 1, 0, 1, 1);
    toolOutputLayout->addWidget(toolUnitAftWid, 2, 0, 1, 1);
    toolOutputLayout->addWidget(toolLabHis, 0, 1, 1, 1);
    toolOutputLayout->addWidget(toolLabBef, 1, 1, 1, 1);
    toolOutputLayout->addWidget(toolLabAft, 2, 1, 1, 1);

    toolOutputLayout->setSpacing(2);
    toolOutputLayout->setContentsMargins(1, 1, 1, 0);

    this->setLayout(toolOutputLayout);
    this->setFixedHeight(270);
    this->setStyleSheet("background-color:#8C8C8C;border-radius:4px;");
}

void ToolModelOutput::initUnitList(QString listStr)
{
    QStringList unitStrList = listStr.split(",");

    QListWidgetItem *unitBeforeItem[unitStrList.size()];
    QListWidgetItem *unitAfterItem[unitStrList.size()];

    unitListBef->clear();
    unitListAft->clear();

    // 单位列表初始化
    for (int i = 0; i < unitStrList.size(); i++) {

        unitBeforeItem[i] = new QListWidgetItem(unitStrList[i], unitListBef);
        unitAfterItem[i]  = new QListWidgetItem(unitStrList[i], unitListAft);

        unitListBef->addItem(unitBeforeItem[i]);
        unitListAft->addItem(unitAfterItem[i]);
    }

    unitListBef->setSpacing(2);
    unitListAft->setSpacing(2);

    QString unitListStyle = "QListWidget::item:selected{background-color:#36363A;}"
                            "QListWidget::item:hover{background-color:#666666;}";

    unitListBef->setStyleSheet(unitListStyle);
    unitListAft->setStyleSheet(unitListStyle);
}

void ToolModelOutput::unitConversion()
{
    QString strBefore = this->toolLabBef->text();
    QString strAfter  = this->toolLabAft->text();
}

ToolModelButton::ToolModelButton(QWidget *parent)
    : QWidget(parent)
{
    //布局按钮
    QFont btnFont("HelveticaNeueLTPro-UltLt", 30, 10);

    QString btnStyle = "QPushButton{"
                       "width:106px;height:62px;opacity:0.95;"
                       "background-color:#1C1C1E;border-radius:4px;"
                       "font-size:42px;font-family:HelveticaNeueLTPro-UltLt;color:#FFFFFF;"
                       "}"
                       "QPushButton:hover{background-color:rgb(180,180,180);}";

    for (int i = 0; i < 10; i++) {
        btnNum[i] = new QPushButton(this);
        btnNum[i]->setText(QString::number(i));
        btnNum[i]->setStyleSheet(btnStyle);
        btnNum[i]->setFont(btnFont);
        btnNum[i]->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Expanding);
    }

    QString btnOpStyle = "QPushButton{"
                         "width:106px;height:62px;opacity:1;"
                         "background-color:#222E36;border-radius:4px;"
                         "font-size:35px;font-family:HelveticaNeueLTPro-UltLt;color:#FFFFFF;}"
                         "QPushButton:hover{background-color:rgba(82,87,217,1);}";

    QString btnAddStyle = "QPushButton{"
                          "width:106px;height:62px;opacity:1;"
                          "background-color:#222E36;border-radius:4px;"
                          "font-size:50px;font-family:HelveticaNeueLTPro-UltLt;color:#FFFFFF;"
                          "}"
                          "QPushButton:hover{background-color:rgba(82,87,217,1);}";

    QString btnEqualStyle = "QPushButton{"
                            "width:106px;height:126px;opacity:1;"
                            "background-color:#E64A19;border-radius:4px;"
                            "font-size:48px;font-family:HelveticaNeueLTPro-UltLt;color:#FFFFFF;"
                            "}"
                            "QPushButton:hover{background-color:rgba(82,87,217,1);}";
    QString btnPerStyle = "QPushButton{"
                          "width:106px;height:62px;opacity:1;"
                          "background-color:#1C1C1E;border-radius:4px;"
                          "font-size:31px;font-family:HelveticaNeueLTPro-UltLt;color:#FFFFFF;"
                          "}"
                          "QPushButton:hover{background-color:rgb(180,180,180);}";

    btnClear = new QPushButton(this);
    btnDiv = new QPushButton(this);
    btnMulti = new QPushButton(this);
    btnDelete = new QPushButton(this);
    btnSub = new QPushButton(this);
    btnAdd = new QPushButton(this);
    btnEqual = new QPushButton(this);
    btnPer = new QPushButton(this);
    btnPoint = new QPushButton(this);

    // 设置按钮的显示文本
    QString btnList = "C,÷,×,,—,+,=,%,.";
    QStringList btnNameList= btnList.split(",");
    int index = 0;

    btnClear->setText(btnNameList[index++]);
    btnDiv->setText(btnNameList[index++]);
    btnMulti->setText(btnNameList[index++]);
    btnDelete->setText(btnNameList[index++]);
    btnSub->setText(btnNameList[index++]);
    btnAdd->setText(btnNameList[index++]);
    btnEqual->setText(btnNameList[index++]);
    btnPer->setText(btnNameList[index++]);
    btnPoint->setText(btnNameList[index++]);

    // 设置按钮字体
    btnClear->setFont(btnFont);
    btnDiv->setFont(btnFont);
    btnMulti->setFont(btnFont);
    btnDelete->setFont(btnFont);
    btnSub->setFont(btnFont);
    btnAdd->setFont(btnFont);
    btnEqual->setFont(btnFont);
    btnPer->setFont(btnFont);
    btnPoint->setFont(btnFont);

    // 设置按钮样式standardModel
    btnClear->setStyleSheet(btnOpStyle);
    btnDiv->setStyleSheet(btnOpStyle);
    btnMulti->setStyleSheet(btnOpStyle);
    btnDelete->setStyleSheet(btnOpStyle);
    btnSub->setStyleSheet(btnOpStyle);
    btnAdd->setStyleSheet(btnAddStyle);
    btnEqual->setStyleSheet(btnEqualStyle);
    btnPer->setStyleSheet(btnPerStyle);
    btnPoint->setStyleSheet(btnPerStyle);

    // 设置按钮自适应放缩
    btnClear->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Expanding);
    btnDiv->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Expanding);
    btnMulti->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Expanding);
    btnDelete->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Expanding);
    btnSub->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Expanding);
    btnAdd->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Expanding);
    btnEqual->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Expanding);
    btnPer->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Expanding);
    btnPoint->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Expanding);

    // 删除按钮
    btnDelete->setIcon(QIcon(":/image/delete.png"));
    btnDelete->setIconSize(QSize(60, 40));

    // 将按钮进行网格布局
    QGridLayout *btnLayout = new QGridLayout();
    btnLayout->addWidget(btnClear, 0, 0, 1, 1);
    btnLayout->addWidget(btnDiv, 0, 1, 1, 1);
    btnLayout->addWidget(btnMulti, 0, 2, 1, 1);
    btnLayout->addWidget(btnDelete, 0, 3, 1, 1);
    btnLayout->addWidget(btnSub, 1, 3, 1, 1);
    btnLayout->addWidget(btnAdd, 2, 3, 1, 1);
    btnLayout->addWidget(btnEqual, 3, 3, 2, 1);
    btnLayout->addWidget(btnPer, 4, 0, 1, 1);
    btnLayout->addWidget(btnPoint, 4, 2, 1, 1);

    // 数字按钮布局
    btnLayout->addWidget(btnNum[0], 4, 1, 1, 1);
    for (int i = 1; i < 10; i++) {
        btnLayout->addWidget(btnNum[i], 3 - (i - 1)/3, (i - 1) % 3, 1, 1);
    }

    // 设置间距和背景样式
    btnLayout->setSpacing(2);
    btnLayout->setMargin(1);

    this->setFixedHeight(320);
    this->setLayout(btnLayout);
    this->setStyleSheet("border-radius:4px;");
}

void ToolModelButton::keyPressEvent(QKeyEvent *event)
{
    switch (event->key())
    {
    case Qt::Key_Period:
        btnPoint->click();
        break;
    case Qt::Key_0:
        btnNum[0]->click();
        break;
    case Qt::Key_1:
        btnNum[1]->click();
        break;
    case Qt::Key_2:
        btnNum[2]->click();
        break;
    case Qt::Key_3:
        btnNum[3]->click();
        break;
    case Qt::Key_4:
        btnNum[4]->click();
        break;
    case Qt::Key_5:
        btnNum[5]->click();
        break;
    case Qt::Key_6:
        btnNum[6]->click();
        break;
    case Qt::Key_7:
        btnNum[7]->click();
        break;
    case Qt::Key_8:
        btnNum[8]->click();
        break;
    case Qt::Key_9:
        btnNum[9]->click();
        break;
    case Qt::Key_Plus:
        btnAdd->click();
        break;
    case Qt::Key_Minus:
        btnSub->click();
        break;
    case Qt::Key_Asterisk:
        btnMulti->click();
        break;
    case Qt::Key_Slash:
        btnDiv->click();
        break;
    case Qt::Key_Escape:
        btnClear->click();
        break;
    case Qt::Key_Percent:
        btnPer->click();
        break;
    case Qt::Key_Backspace:
        btnDelete->click();
        break;
    case Qt::Key_Enter:
        btnEqual->click();
        break;
    case 16777220:
        btnEqual->click();
        break;
    case 0x3d:
        btnEqual->click();
        break;
    default:
        break;
    }
}
